sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSchemaInferenceSuite.scala

```diff
@@ -0,0 +1,352 @@
+import org.apache.spark.sql.internal.SQLConf.HiveCaseSensitiveInferenceMode.{
+  Value => InferenceMode, _}
+import org.apache.spark.sql.test.SQLTestUtils
+import org.apache.spark.sql.types._
+
+class HiveSchemaInferenceSuite
+  extends QueryTest with TestHiveSingleton with SQLTestUtils with BeforeAndAfterEach {
+
+  import HiveSchemaInferenceSuite._
+  import HiveExternalCatalog.SPARK_SQL_PREFIX
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    FileStatusCache.resetForTesting()
+  }
+
+  override def afterEach(): Unit = {
+    super.afterEach()
+    spark.sessionState.catalog.tableRelationCache.invalidateAll()
+    FileStatusCache.resetForTesting()
+  }
+
+  private val externalCatalog = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog]
+  private val client = externalCatalog.client
+
+  // Return a copy of the given schema with all field names converted to lower case.
+  private def lowerCaseSchema(schema: StructType): StructType
+    = StructType(schema.map(f => f.copy(name = f.name.toLowerCase)))
+
+  // Create a CatalogTable instance modeling an external Hive Metastore table backed by
+  // Parquet data files.
+  private def hiveExternalCatalogTable(
+      tableName: String,
+      serde: HiveSerDe,
+      location: String,
+      schema: StructType,
+      partitionColumns: Seq[String],
+      properties: Map[String, String] = Map.empty): CatalogTable = {
+    CatalogTable(
+      identifier = TableIdentifier(table = tableName, database = Option(DATABASE)),
+      tableType = CatalogTableType.EXTERNAL,
+      storage = CatalogStorageFormat(
+        locationUri = Option(new java.net.URI(location)),
+        inputFormat = serde.inputFormat,
+        outputFormat = serde.outputFormat,
+        serde = serde.serde,
+        compressed = false,
+        properties = Map("serialization.format" -> "1")),
+      schema = schema,
+      provider = Option("hive"),
+      partitionColumnNames = partitionColumns,
+      properties = properties)
+  }
+
+  // Create a CatalogTablePartition instance for adding partitions of data to our test table.
+  private def hiveCatalogPartition(
+      location: String,
+      serde: HiveSerDe,
+      partitionCols: Seq[String],
+      index: Int): CatalogTablePartition = {
+    val locationUri =
+      new java.net.URI(location + "/" +  partitionCols.map(c => s"$c=$index").mkString("/") + "/")
+    CatalogTablePartition(
+      spec = partitionCols.map(_.toLowerCase -> index.toString).toMap,
+      storage = CatalogStorageFormat(
+        locationUri = Option(locationUri),
+        inputFormat = serde.inputFormat,
+        outputFormat = serde.outputFormat,
+        serde = serde.serde,
+        compressed = false,
+        properties = Map("serialization.format" -> "1")))
+  }
+
+  // Create a Hive external test table containing the given field and partition column names.
+  // Returns a case-sensitive schema for the table.
+  private def setupExternalTable(
```

Comment:
```fileType``` is not fixed-- it is either ```"orc"``` or ```"parquet"``` (the test now covers both file formats). ```tableName``` could be made a constant I suppose since the table is added and removed each time. The other fields should all vary per test.